torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -3078,25 +3099,49 @@ class ForceFP16Annotation : public IterVisitor {
     return annotation.force_fp16_tv_set_;
   }
 
+ public:
+  static std::unordered_set<TensorView*> getBF16AnnotatedSet(Fusion* fusion) {
+    ForceHalfAnnotation annotation;
+    std::vector<Val*> bf16_outputs;
+
+    std::copy_if(
+        fusion->outputs().begin(),
+        fusion->outputs().end(),
+        std::back_inserter(bf16_outputs),
+        [](auto* val) {
+          return val->template isA<TensorView>() &&
+              val->getDataType().has_value() &&
+              val->getDataType().value() == DataType::BFloat16;
+        });
+
+    annotation.traverseFrom(fusion, bf16_outputs);
+    return annotation.force_bf16_tv_set_;
+  }
+
  private:
   using IterVisitor::handle;
 
   void handle(TensorView* tv) override {
     auto dtype = tv->getDataType();
     if (dtype.has_value() && dtype.value() == DataType::Float &&
-        !tv->isFusionOutput() && !tv->isFusionInput()) {
+        !tv->isFusionOutput() && !tv->isFusionInput() && tv->getDataType() == DataType::Half) {
       force_fp16_tv_set_.insert(tv);
     }
+    else if (dtype.has_value() && dtype.value() == DataType::Float &&
```

Comment:
I guess it'd be cleaner to just check `tv->getDataType() == DataType::Half || tv->getDataType() == DataType::BFloat16` instead of having another branch that's almost the same.

Reply:
Also please feel free to add a flag in this class, i.e `DataType cast_to_type` which is bt16 or half and add an assertion here to ensure we don't see a mixed bf16 and half graph.